Mainnet/AddressMapper/AddressMapper.Tests/AddressMapperTests.cs

```diff
@@ -0,0 +1,145 @@
+using Moq;
+using Stratis.SmartContracts.CLR;
+using Stratis.SmartContracts;
+using Xunit;
+using static AddressMapper;
+
+namespace Tests
+{
+    public class AddressMapperTests
+    {
+        private readonly Mock<ISmartContractState> mContractState;
+        private readonly Mock<IContractLogger> mContractLogger;
+        private readonly Mock<IInternalTransactionExecutor> mInternalExecutor;
+
+        private readonly IPersistentState persistentState;
+
+        private readonly Address contractAddress;
+        private readonly Address primaryAddress;
+        private readonly Address secondaryAddress;
+        private readonly Address ownerAddress;
+
+        public AddressMapperTests()
+        {
+            mContractLogger = new Mock<IContractLogger>();
+            persistentState = new InMemoryState();
+            mContractState = new Mock<ISmartContractState>();
+            mInternalExecutor = new Mock<IInternalTransactionExecutor>();
+            mContractState.Setup(x => x.PersistentState).Returns(persistentState);
+            mContractState.Setup(x => x.ContractLogger).Returns(mContractLogger.Object);
+            mContractState.Setup(x => x.InternalTransactionExecutor).Returns(mInternalExecutor.Object);
+            contractAddress = "0x0000000000000000000000000000000000000001".HexToAddress();
+            ownerAddress = "0x0000000000000000000000000000000000000002".HexToAddress();
+            primaryAddress = "0x0000000000000000000000000000000000000003".HexToAddress();
+            secondaryAddress = "0x0000000000000000000000000000000000000004".HexToAddress();
+        }
+
+        [Theory]
+        [InlineData(Status.Approved)]
+        [InlineData(Status.Pending)]
+        public void MappAddress_Fails_If_SecondaryAddress_Is_Mapped_Already(Status status)
+        {
+            persistentState.SetStruct($"MappingInfo:{secondaryAddress}", new MappingInfo { Primary = primaryAddress, Status = (int)status });
+
+            mContractState.Setup(x => x.Message).Returns(new Message(contractAddress, primaryAddress, 0));
+
+            var contract = new AddressMapper(mContractState.Object, ownerAddress);
+
+            Assert.Throws<SmartContractAssertException>(() => contract.MapAddress(secondaryAddress));
+        }
+
+        [Fact]
+        public void MappAddress_Maps_PrimaryAddres()
+        {
+            persistentState.SetStruct($"MappingInfo:{secondaryAddress}", new MappingInfo { Primary = primaryAddress, Status = (int)Status.NoStatus });
+
+            mContractState.Setup(x => x.Message).Returns(new Message(contractAddress, primaryAddress, 0));
+
+            var contract = new AddressMapper(mContractState.Object, ownerAddress);
+
+            contract.MapAddress(secondaryAddress);
+
+            Assert.Equal(new MappingInfo { Primary = primaryAddress, Status = (int)Status.Pending }, persistentState.GetStruct<MappingInfo>($"MappingInfo:{secondaryAddress}"));
```

Comment:
Because you set the struct on Line 54 above, line 62 is always going to be true. 

It would be better to use 
```
mContractState.Verify(x => x.SetStruct<MappingInfo>(...), Times.Once);
```

to validate that the contract call actually made it there in code and persisted the correct values on line 61 after you call the contract method.

Reply:
Line 54 sets `NoStatus` for `MappingInfo` and line 62 is verify that it is set to `Pending` state after `MappAddress`called. 

I just realized that line 54 is unnecessary. I removed it. 

I can not call verify method because persistentState is not a mocked object. It is InMemoryState for easier testing.